### PR TITLE
Cancel compute plan

### PR DESCRIPTION
**Pull request by inelgnu**
Companion PR:
- [substra](https://github.com/SubstraFoundation/substra/pull/56)
- [backend](https://github.com/SubstraFoundation/substra-backend/pull/65)
- [tests](https://github.com/SubstraFoundation/substra-tests/pull/26)
- [frontend](https://github.com/SubstraFoundation/substra-frontend/pull/33)

### CODE REVIEW

File: chaincode/utils.go [hunk 1]
```diff
@@ -155,3 +155,13 @@ func (modelType CompositeModelType) String() string {
...
+
+// Contains check if an element is contained in a slice of string
+func Contains(a []string, x string) bool {
+	for _, s := range a {
+		if s == x {
+			return true
+		}
+	}
+	return false
+}
```
Comment (thibaultrobert) [inline chaincode/utils.go]: There is already a function `stringInSlice` in `utils.go` that do exactly that.

File: chaincode/compute_plan.go [hunk 2]
```diff
@@ -311,13 +311,62 @@ func getComputePlan(db *LedgerDB, key string) (resp outputComputePlan, err error
...
 	}
 	return
 }
+
+func getComputePlanStatus(db *LedgerDB, aggregatetupleKeys, traintupleKeys, compositeTraintupleKeys []string) (status string, err error) {
+	// get all tuples like status in one slice
+	statusCollection := []string{}
+
+	for _, aggregatetupleKey := range aggregatetupleKeys {
+		aggregatetuple, err := getOutputAggregatetuple(db, aggregatetupleKey)
```
Comment (thibaultrobert) [inline chaincode/compute_plan.go]: Why getting the output tuples here? If there is no specific need you should only get the intern object (here it will be `db.GetAggregatetuple`). More over it exists a function `db.GetGenericTuple` that work for every tuple and allow you to acces common part of tuples including the status.

Comment (inelgnu) [reply chaincode/compute_plan.go]: fixed here ef97caf

File: chaincode/compute_plan.go [hunk 3]
```diff
@@ -311,13 +311,45 @@ func getComputePlan(db *LedgerDB, key string) (resp outputComputePlan, err error
 		}
 	}
 
+
+	status, _ := getComputePlanStatus(db, tupleKeys)
```
Comment (thibaultrobert) [inline chaincode/compute_plan.go]: We should check the error here even for the status computation. Also I think it would be good to return the compute plan at creation.

**Comment by AurelienGasser** [comment]
@inalgnu Sorry about the duplicate work ! (https://github.com/SubstraFoundation/substra-chaincode/pull/52)

Feel free to cherry pick some changes from there. I think the only valuable things to potentially bring over are 

- [these changes](https://github.com/SubstraFoundation/substra-chaincode/pull/52/files#diff-f0db4dc81b7e8c20cb3ecbea544ced0bR161-R293) to the composite/aggregate workflow test. 
- maybe also [these extra test cases](https://github.com/SubstraFoundation/substra-chaincode/pull/52/files#diff-f0db4dc81b7e8c20cb3ecbea544ced0bR500-R543)

Note that in https://github.com/SubstraFoundation/substra-chaincode/pull/52, and unlike this PR, we don't fetch _all_ the tuples from the ledger before determining the status of the compute plan. The _only_ advantage to doing that is for the "Canceled" case: we can fetch only one tuple and immediately return that the compute plan is "Canceled", without fetching any of the other tuples. I think this small optimization is not worth it, considering it makes the code (including tests) more complicated. Do you agree ? cc @thibaultrobert

**Comment by thibaultrobert** [comment]
> Note that in #52, and unlike this PR, we don't fetch _all_ the tuples from the ledger before determining the status of the compute plan. The _only_ advantage to doing that is for the "Canceled" case: we can fetch only one tuple and immediately return that the compute plan is "Canceled", without fetching any of the other tuples. I think this small optimization is not worth it, considering it makes the code (including tests) more complicated. Do you agree ?

Like you, I tend to think it would be too costly a optimization even though it would reduce the read-set in the `queryComputePlan` smartcontract. And even if it was a bit complex to read, I did enjoy the function injection in your solution.

**Comment by thibaultrobert** [comment]
I took upon myself to update the `README.md` to fix the tests @inalgnu. And I will also set the compute plan status in the `createComputePlan` smartcontract. Fill free to revert if you disagree with any of my commits.

**Comment by thibaultrobert** [comment]
I have a question on the compute plan status:
The status of a compute plan is only compute (yes I just did that) from the train/composite/aggregate tuples and not from the testuples. So a compute plan can be `Done` even if one of the testtuples is `Doing` or even `Failed`.  Is that what was intended? (I didn't find the answer in the [spec](https://github.com/owkin/owkin-connect-spec/blob/master/motiv_modif/fail_fast.md), sorry if I missed it)

**Comment by inelgnu** [comment]
> I have a question on the compute plan status:
> The status of a compute plan is only compute (yes I just did that) from the train/composite/aggregate tuples and not from the testuples. So a compute plan can be `Done` even if one of the testtuples is `Doing` or even `Failed`. Is that what was intended? (I didn't find the answer in the [spec](https://github.com/owkin/owkin-connect-spec/blob/master/motiv_modif/fail_fast.md), sorry if I missed it)

You're right, fixed here 94b0213

File: chaincode/testtuple.go [hunk 4]
```diff
@@ -264,21 +266,35 @@ func logStartTest(db *LedgerDB, args []string) (outputTesttuple outputTesttuple,
...
+	tuple, err := db.GetGenericTuple(testtuple.TraintupleKey)
+	if err != nil {
+		return outputTesttuple{}, err
+	}
+
+	// cancel testtuple if compute plan is canceled
+	if tuple.ComputePlanID != "" {
+		err := cancelIfComputePlanIsCanceled(db, inp.Key, tuple.ComputePlanID, &testtuple)
+		if err != nil {
+			return outputTesttuple{}, err
```
Comment (AurelienGasser) [inline chaincode/testtuple.go:284]: I think you can just do `return` without arguments here

File: chaincode/testtuple.go [hunk 5]
```diff
@@ -264,21 +266,35 @@ func logStartTest(db *LedgerDB, args []string) (outputTesttuple outputTesttuple,
 	if err != nil {
 		return
 	}
+
+	tuple, err := db.GetGenericTuple(testtuple.TraintupleKey)
+	if err != nil {
+		return outputTesttuple{}, err
```
Comment (AurelienGasser) [inline chaincode/testtuple.go:277]: I think you can just do `return` without arguments here

File: chaincode/testtuple.go [hunk 6]
```diff
@@ -290,6 +306,19 @@ func logSuccessTest(db *LedgerDB, args []string) (outputTesttuple outputTesttupl
 		return
 	}
 
+	tuple, err := db.GetGenericTuple(testtuple.TraintupleKey)
+	if err != nil {
+		return outputTesttuple{}, err
```
Comment (AurelienGasser) [inline chaincode/testtuple.go:320]: I think you can just do `return` without arguments here

File: chaincode/testtuple.go [hunk 7]
```diff
@@ -290,6 +306,19 @@ func logSuccessTest(db *LedgerDB, args []string) (outputTesttuple outputTesttupl
...
+	}
+
+	// cancel testtuple if compute plan is canceled
+	if tuple.ComputePlanID != "" {
+		err := cancelIfComputePlanIsCanceled(db, inp.Key, tuple.ComputePlanID, &testtuple)
+		if err != nil {
+			return outputTesttuple{}, err
```
Comment (AurelienGasser) [inline chaincode/testtuple.go:327]: I think you can just do `return` without arguments here

File: chaincode/testtuple.go [hunk 8]
```diff
@@ -317,6 +346,19 @@ func logFailTest(db *LedgerDB, args []string) (outputTesttuple outputTesttuple,
 		return
 	}
 
+	tuple, err := db.GetGenericTuple(testtuple.TraintupleKey)
+	if err != nil {
+		return outputTesttuple{}, err
```
Comment (AurelienGasser) [inline chaincode/testtuple.go:364]: I think you can just do `return` without arguments here

File: chaincode/testtuple.go [hunk 9]
```diff
@@ -317,6 +346,19 @@ func logFailTest(db *LedgerDB, args []string) (outputTesttuple outputTesttuple,
...
+	}
+
+	// cancel testtuple if compute plan is canceled
+	if tuple.ComputePlanID != "" {
+		err := cancelIfComputePlanIsCanceled(db, inp.Key, tuple.ComputePlanID, &testtuple)
+		if err != nil {
+			return outputTesttuple{}, err
```
Comment (AurelienGasser) [inline chaincode/testtuple.go:371]: I think you can just do `return` without arguments here

File: chaincode/traintuple.go [hunk 10]
```diff
@@ -280,19 +280,28 @@ func logStartTrain(db *LedgerDB, args []string) (outputTraintuple outputTraintup
 	if err != nil {
 		return
 	}
+
+	// cancel traintuple if compute plan is canceled
+	if traintuple.ComputePlanID != "" {
+		err := cancelIfComputePlanIsCanceled(db, inp.Key, traintuple.ComputePlanID, &traintuple)
+		if err != nil {
+			return outputTraintuple{}, err
```
Comment (AurelienGasser) [inline chaincode/traintuple.go:279]: I think you can just do `return` without arguments here

File: chaincode/traintuple.go [hunk 11]
```diff
@@ -305,6 +314,14 @@ func logSuccessTrain(db *LedgerDB, args []string) (outputTraintuple outputTraint
 	if err != nil {
 		return
 	}
+	// cancel traintuple if compute plan is canceled
+	if traintuple.ComputePlanID != "" {
+		err := cancelIfComputePlanIsCanceled(db, inp.Key, traintuple.ComputePlanID, &traintuple)
+		if err != nil {
+			return outputTraintuple{}, err
```
Comment (AurelienGasser) [inline chaincode/traintuple.go:316]: I think you can just do `return` without arguments here

File: chaincode/traintuple.go [hunk 12]
```diff
@@ -350,6 +367,14 @@ func logFailTrain(db *LedgerDB, args []string) (outputTraintuple outputTraintupl
...
```
Comment (AurelienGasser) [inline chaincode/traintuple.go:372]: I think you can just do `return` without arguments here

File: chaincode/traintuple_composite.go [hunk 13]
```diff
@@ -301,105 +301,132 @@ func createCompositeTraintupleInternal(db *LedgerDB, inp inputCompositeTraintupl
...
 	if err != nil {
 		return
 	}
-	if err = validateTupleOwner(db, traintuple.Dataset.Worker); err != nil {
+
+	// cancel compositeTraintuple if compute plan is canceled
+	if compositeTraintuple.ComputePlanID != "" {
+		err := cancelIfComputePlanIsCanceled(db, inp.Key, compositeTraintuple.ComputePlanID, &compositeTraintuple)
+		if err != nil {
+			return outputCompositeTraintuple{}, err
```
Comment (AurelienGasser) [inline chaincode/traintuple_composite.go:312]: I think you can just do `return` without arguments here

File: chaincode/traintuple_composite.go [hunk 14]
```diff
@@ -301,105 +301,132 @@ func createCompositeTraintupleInternal(db *LedgerDB, inp inputCompositeTraintupl
...
 		return
 	}
-	traintuple.Perf = inp.Perf
-	traintuple.OutHeadModel.OutModel = &HashDress{
+
+	// cancel compositeTraintuple if compute plan is canceled
+	if compositeTraintuple.ComputePlanID != "" {
+		err := cancelIfComputePlanIsCanceled(db, inp.Key, compositeTraintuple.ComputePlanID, &compositeTraintuple)
+		if err != nil {
+			return outputCompositeTraintuple{}, err
```
Comment (AurelienGasser) [inline chaincode/traintuple_composite.go:350]: I think you can just do `return` without arguments here

File: chaincode/traintuple_composite.go [hunk 15]
```diff
@@ -301,105 +301,132 @@ func createCompositeTraintupleInternal(db *LedgerDB, inp inputCompositeTraintupl
...
 		return
 	}
-	traintuple.Log += inp.Log
 
-	if err = validateTupleOwner(db, traintuple.Dataset.Worker); err != nil {
+	// cancel compositeTraintuple if compute plan is canceled
+	if compositeTraintuple.ComputePlanID != "" {
+		err := cancelIfComputePlanIsCanceled(db, inp.Key, compositeTraintuple.ComputePlanID, &compositeTraintuple)
+		if err != nil {
+			return outputCompositeTraintuple{}, err
```
Comment (AurelienGasser) [inline chaincode/traintuple_composite.go:406]: I think you can just do `return` without arguments here

File: chaincode/tuple_aggregate.go [hunk 16]
```diff
@@ -295,18 +295,27 @@ func logStartAggregate(db *LedgerDB, args []string) (outputAggregatetuple output
 	if err != nil {
 		return
 	}
+
+	// cancel aggregatetuple if compute plan is canceled
+	if aggregatetuple.ComputePlanID != "" {
+		err := cancelIfComputePlanIsCanceled(db, inp.Key, aggregatetuple.ComputePlanID, &aggregatetuple)
+		if err != nil {
+			return outputAggregatetuple{}, err
```
Comment (AurelienGasser) [inline chaincode/tuple_aggregate.go:285]: I think you can just do `return` without arguments here

File: chaincode/tuple_aggregate.go [hunk 17]
```diff
@@ -318,6 +327,15 @@ func logFailAggregate(db *LedgerDB, args []string) (outputAggregatetuple outputA
...
```
Comment (AurelienGasser) [inline chaincode/tuple_aggregate.go:321]: ok that's wayyy to many times I copy/pasted my comment already :D I'll stop now.

Comment (thibaultrobert) [reply chaincode/tuple_aggregate.go:321]: It would work at every time you copy/pasted it indeed but I think it's intentional here. @inalgnu will have to confirm it but return without explicit arguments are considered confusing and to be avoid.
Named return values are for documentation mostly but even then you should specified the return arguments. Here it would be `return o, err` at the very least.
It's mostly my bad if you've find a lot of bare return around the base code. Sorry about that.

Comment (AurelienGasser) [reply chaincode/tuple_aggregate.go:321]: Interesting, thanks for the precision! Let's see what @inalgnu says.

Comment (inelgnu) [reply chaincode/tuple_aggregate.go:321]: @AurelienGasser named params could be easy to use but might cause some issues especially shadowing. I used to avoid using them as a good practice.
this won't compile for example => https://play.golang.org/p/2rumK4hdKec
from codereviewcomments:
https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters

> Don't name result parameters just to avoid declaring a var inside the function; that trades off a minor implementation brevity at the cost of unnecessary API verbosity.

In addition, here outputAggregatetuple type was named same as the parameter name which leads to an error if you want to create a new struct outputAggregatetuple{}
https://play.golang.org/p/g0RuGAR6W9t

Erratum:
Here is an example of shadowing, thank's to @thibaultrobert feedback on the first example which doesn't compile and it's a good news 😊 
https://play.golang.org/p/KL0mx9pvOo1

**Comment by inelgnu** [comment]
@AurelienGasser  Sorry for the delay 😞 

> @inalgnu Sorry about the duplicate work ! (#52)
> 
> Feel free to cherry pick some changes from there. I think the only valuable things to potentially bring over are
> 
> * [these changes](https://github.com/SubstraFoundation/substra-chaincode/pull/52/files#diff-f0db4dc81b7e8c20cb3ecbea544ced0bR161-R293) to the composite/aggregate workflow test.
> * maybe also [these extra test cases](https://github.com/SubstraFoundation/substra-chaincode/pull/52/files#diff-f0db4dc81b7e8c20cb3ecbea544ced0bR500-R543)
> 
@AurelienGasser i would cherry-pick the tests with pleasure. Could you please separate them in atomic commits, or just commit them directly on this PR. Thank's :-) 

> Note that in #52, and unlike this PR, we don't fetch _all_ the tuples from the ledger before determining the status of the compute plan. The _only_ advantage to doing that is for the "Canceled" case: we can fetch only one tuple and immediately return that the compute plan is "Canceled", without fetching any of the other tuples. I think this small optimization is not worth it, considering it makes the code (including tests) more complicated. Do you agree ? cc @thibaultrobert

Your approach is way better, however, I'm not sure it worth it, for now, maybe we should measure before optimizing. What do you think @thibaultrobert ?

**Comment by thibaultrobert** [comment]
> Your approach is way better, however, I'm not sure it worth it, for now, maybe we should measure before optimizing. What do you think @thibaultrobert ?

Like I answer before, I don't think this optimization is worth its load for now. We should measure first as you said.

File: chaincode/compute_plan.go [hunk 18]
```diff
@@ -220,6 +224,7 @@ func createComputePlanInternal(db *LedgerDB, inp inputComputePlan) (resp outputC
 		resp.TesttupleKeys = append(resp.TesttupleKeys, testtupleKey)
```
Comment (thibaultrobert) [inline chaincode/compute_plan.go:217]: We should add the testtuples' keys here I think.

File: chaincode/compute_plan.go [hunk 19]
```diff
@@ -311,13 +316,116 @@ func getComputePlan(db *LedgerDB, key string) (resp outputComputePlan, err error
...
+		return "", err
+	}
+
+	keys := []string{}
+	keys = append(keys, computePlan.TraintupleKeys...)
+	keys = append(keys, computePlan.CompositeTraintupleKeys...)
+	keys = append(keys, computePlan.AggregatetupleKeys...)
+	keys = append(keys, computePlan.TesttupleKeys...)
+
+	return getComputePlanStatus(db, keys)
```
Comment (thibaultrobert) [inline chaincode/compute_plan.go]: Now that I've read it like that I tend to think that `getComputePlan` should perhaps take the compute plan directly as argument and we should reconstruct this list in the function. What do you think @inalgnu

File: chaincode/compute_plan.go (hunk 18 above)
Comment (thibaultrobert) [reply chaincode/compute_plan.go:217]: After a quick exchange with @inalgnu , in the case of the `createComputePlan`, all the testtuples are supposed to be in a `waitingStatus` so there are not relevant to determine the compute plan global status. 
Sorry for the noise.

File: chaincode/compute_plan.go (hunk 19 above)
Comment (inelgnu) [reply chaincode/compute_plan.go]: @thibaultrobert that make sens, updated

File: chaincode/compute_plan.go [hunk 20]
```diff
@@ -319,5 +320,104 @@ func getComputePlan(db *LedgerDB, key string) (resp outputComputePlan, err error
...
+		if err != nil {
+			return false, err
+		}
+		err = testtuple.commitStatusUpdate(db, key, StatusCanceled)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	return true, nil
```
Comment (thibaultrobert) [inline chaincode/compute_plan.go]: Like I said directly to @inalgnu, the return value of a smartcontract should respect the [interface `Unmarshaler`](https://golang.org/pkg/encoding/json/#Unmarshaler).

File: chaincode/ledger.go [hunk 21]
```diff
@@ -37,6 +37,11 @@ const (
 	// the String() function in utils.go
 )
 
+// StatusUpdater is exported
+type StatusUpdater interface {
```
Comment (thibaultrobert) [inline chaincode/ledger.go:41]: The power of interface has come among us. 👍

File: chaincode/testtuple.go [hunk 22]
```diff
@@ -159,6 +159,8 @@ func (testtuple *Testtuple) SetFromTraintuple(db *LedgerDB, traintupleKey string
 		return errors.BadRequest(
 			"could not register this testtuple, the traintuple %s has a failed status",
 			traintupleKey)
```
Comment (thibaultrobert) [inline chaincode/testtuple.go:165]: That has nothing to do with this PR per say but we are updating the different train like tuples to be created in a `StatusFailed` if their inModels are failed. Do we want a similar approach for testtuples? (Here we return an error, we could create the testtuple in a failed state)

File: chaincode/testtuple.go [hunk 23]
```diff
@@ -264,21 +266,35 @@ func logStartTest(db *LedgerDB, args []string) (outputTesttuple outputTesttuple,
...
 	}
+
+	tuple, err := db.GetGenericTuple(testtuple.TraintupleKey)
+	if err != nil {
+		return outputTesttuple{}, err
+	}
+
+	// cancel testtuple if compute plan is canceled
+	if tuple.ComputePlanID != "" {
+		err := cancelIfComputePlanIsCanceled(db, inp.Key, tuple.ComputePlanID, &testtuple)
```
Comment (thibaultrobert) [inline chaincode/testtuple.go]: I think there is a problem here: we should return after canceling this tuple. If we continue, we will update the status later in the function to `StatusDoing` or am I missing something obvious?
And this apply to every other `LogXXX` function.

File: chaincode/compute_plan.go (hunk 20 above)
Comment (inelgnu) [reply chaincode/compute_plan.go]: done, thank's

File: chaincode/tuple_aggregate.go (hunk 17 above)
Comment (AurelienGasser) [reply chaincode/tuple_aggregate.go:321]: Thank you, and sorry for the avalanche of comments!

File: chaincode/testtuple.go (hunk 23 above)
Comment (inelgnu) [reply chaincode/testtuple.go]: yes, I will fix that

Comment (inelgnu) [reply chaincode/testtuple.go]: fixed here 681506d

**Comment by thibaultrobert** [comment]
I've just made a attempt to keep the interface approach to update the status. The problem with the current approach is that I created a getter on an interface and I'm not 100% comfortable with that.
What do you think?
A partial Update of the JSON is also a possible solution to update status with just `GenericTuple`

**Review by thibaultrobert — APPROVED** [review]

File: chaincode/ledger.go [hunk 24]
```diff
@@ -133,6 +138,7 @@ type CompositeTraintuple struct {
 	InTrunkModel  string                      `json:"inTrunkModel"`
 	OutHeadModel  CompositeTraintupleOutModel `json:"outHeadModel"`
 	OutTrunkModel CompositeTraintupleOutModel `json:"outTrunkModel"`
+	Perf          float32                     `json:"perf"`
```
Comment (AurelienGasser) [inline chaincode/ledger.go:141]: @inalgnu I just noticed this line, is this addition intentional?

Comment (inelgnu) [reply chaincode/ledger.go:141]: nope this seems to be due to a bad rebase :-/

Comment (AurelienGasser) [reply chaincode/ledger.go:141]: https://github.com/SubstraFoundation/substra-chaincode/pull/61
